### PR TITLE
Preserve begin-end keywords delimiting match cases

### LINKED ISSUE #1599: Bug: `exp-grouping=preserve` is not respected when formatting match cases
Formatting code turns `begin`...`end` blocks within match expressions into parenthesized blocks, even when `exp-grouping` is set to `preserve`.

**How to Reproduce**

Version: 0.16.0

In `test.ml`:

```ocaml
type t = Alpha | Beta

let _ = 
  match Alpha with
  | Alpha -> begin
    match Beta with
    | Alpha -> print_endline "Alpha, Alpha"
    | Beta -> print_endline "Alpha, Beta"
  end
  | Beta -> begin
    match Beta with
    | Alpha -> print_endline "Beta, Alpha"
    | Beta -> print_endline "Beta, Beta"
  end
```

Run:

`ocamlformat --enable-outside-detected-project --exp-grouping=preserve test.ml`

Produces:

```ocaml
type t = Alpha | Beta

let _ =
  match Alpha with
  | Alpha -> (
      match Beta with
      | Alpha -> print_endline "Alpha, Alpha"
      | Beta -> print_endline "Alpha, Beta")
  | Beta -> (
      match Beta with
      | Alpha -> print_endline "Beta, Alpha"
      | Beta -> print_endline "Beta, Beta")
```

I've also run into this issue in a project with an `.ocamlformat` file defined. For example, this file:

```
profile = conventional
indicate-multiline-delimiters = closing-on-separate-line
exp-grouping = preserve
```

Will respect the `indicate-multiline-delimiters` argument, but will still convert begin/end to parentheses.

### PR DESCRIPTION
**Pull request by gpetiot**
Fix #1599, no diff with test_branch.sh

We can come up with a cleaner solution when we rewrite the formatting of match cases (these ones depend on so many options!)

### CODE REVIEW

File: test/passing/tests/exp_grouping.ml.ref [hunk 1]
```diff
@@ -272,3 +272,16 @@ let _ =
     foo.fooooo <- Fooo.foo fooo foo.fooooo ;
     Fooo fooo
   end
+
+let _ =
+  match x with
+  | A -> begin
+    match B with A -> fooooooooooooo | B -> fooooooooooooo
+  end
```
Comment (Julow) [inline test/passing/tests/exp_grouping.ml.ref:280]: Would it be rendered as a single line if it fits ? I think we need this test case.

File: lib/Fmt_ast.ml [hunk 2]
```diff
@@ -3050,15 +3053,21 @@ and fmt_case c ctx ~first ~last ~padding case =
                     fmt "@;<1 2>when " $ fmt_expression c (sub_exp ~ctx g) )
               )
           $ p.break_before_arrow $ str "->" $ p.break_after_arrow
-          $ fmt_if parens_here " (" )
+          $ fmt_if parens_here
+              ( match grouping with
+              | `Parens -> " ("
+              | `Begin_end -> "@;<1 0>begin" ) )
```
Comment (Julow) [inline lib/Fmt_ast.ml:3059]: Can't this be moved into `Params.get_cases` ?
Also, that's yet an other piece of code rendering begin/end, I'd prefer if we can find a way to handle that in a consistent way (from `Params` maybe ?).

File: test/passing/tests/exp_grouping.ml.ref (hunk 1 above)
Comment (gpetiot) [reply test/passing/tests/exp_grouping.ml.ref]: It wasn't in my first commit, now it does and I added more tests to see what it looks like for combinations of `break-cases` and `exp-grouping`

File: lib/Fmt_ast.ml (hunk 2 above)
Comment (gpetiot) [reply lib/Fmt_ast.ml]: done!

**Review by Julow — APPROVED** [review]
Looks good.

**Comment by shrik450** [comment]
Thank you very much!

**Pull request merged**